File: client/src/types/Person.js
// @flow

import type { Locateable } from './concerns/Locateable';
import type { Participation } from './Participation';
import type { Qualifiable } from './concerns/Qualifiable';

export type Person = Locateable & Qualifiable & {
  id: number,
  name: string,
  display_name: string,
  person_type: string,
  nationality: string,
  authorized_vocabulary: string,
  url: string,
  database_value: string,
  comment: string,
  part_of: number,
  same_as: number,

  participations: Array<Participation>
};

File: app/models/person.rb
class Person < ApplicationRecord
  # Includes
  include Locateable
  include Qualifiable
  include Recordable
  include Search::Person

  # Relationships
  has_many :participations, dependent: :destroy

  # Nested attributes
  accepts_nested_attributes_for :participations, allow_destroy: :true

  # Resourceable parameters
  allow_params :name, :display_name, :person_type, :authorized_vocabulary, :url, :database_value, :comment, :part_of,
               :same_as, :artist_birth_date, :artist_death_date, :years_active,
               participations_attributes: [:id, :participateable_id, :participateable_type,
                                           :description, :certainty, :notes,
                                           :_destroy, qualifications_attributes: [:id, :value_list_id, :notes, :persistent, :_destroy]]
end
